### DISCUSSION TITLE
Very long segmentation time

### DISCUSSION THREAD

**Discussion by Debo12345678** [neuropoly/axondeepseg#774] [Help needed] [answered]
Hello, I use AxondeepSeg on Windows with the default BF model. I installed it following the steps carefully and only did it once. (My processor is Intel(R) Core(TM) i5-3470 CPU @ 3.20GHz)

When I use good images, segmentation is relatively fast but using napari is very difficult, the application lags when I want to zoom or move the segmented image which is very inconvenient, the latency is about 3 seconds between the moment I zoom with my mouse and the moment the image is zoomed. 

When I use the console to segment a good quality image, segmentation takes less than 5 minutes.  
When segmenting a complex image, the semi-automatic software we use detects axons very poorly because of very clear myelin staining. I wanted to segment the image using a range of zoom factors to find the best images. The segmentation took exactly 10 minutes for the first zoom and 15 minutes for the second zoom, etc..

The command was: 
axondeepseg -t BF -i C:\Users\NFT10\Documents\NewMorpho\ID8.tif -s 0.096 --overlap 60 -r 1.5 3.0 -l 5

And the terminal looked like this:
 
![Capture](https://github.com/axondeepseg/axondeepseg/assets/119101905/45073fa3-f4c3-445b-bdf2-1b85fc0ae6b6)

Are these segmentation times normal?

**Comment by mathieuboudreau** [comment]
@Debo12345678 would you mind sharing what you consider a good quality image vs complex image?

**Reply by mathieuboudreau** [reply]
As for your issue, I can see that your image is being divided into 121 patches of 512x512. What is your original image size and resolution? It's possible that it would perform better if we made it either use less patches, or reduce your image size prior to processing

**Reply by Stoyan-I-A** [reply]
It is possible that the segmentation using napari takes more time because your CPU is tasked with both segmenting the image and running napari. This is also why using napari might feel slower.

Keep also in mind that your CPU, Intel i5-3470, was released in 2012. It will work, but will take more time than newer hardware.

**Reply by Debo12345678** [reply]
Thanks for your quick answers

For pixel size, I used a method that wasn't necessarily the best. The tech lab gave me an image containing the scale bar and told me that the distance between 0 and 10 was 100µm. So I opened this image in imageJ to set scale like this (imageJ know distance is in mm): 

![Set scale](https://github.com/axondeepseg/axondeepseg/assets/119101905/0b22158c-d808-4773-96ab-7a9703762ef6)

Since: 0.1mm <=> 1040.0692pixels
Then 1 pixel <=> 0.00009614mm

So when using AxonDeepSeg, I specify my pixel size as 0,96µm. (The size of the scale bar image is 2382x2706 pixel).  Is there a better way to determine pixel size? 

And the size of my image to be segmented is 3418x3404 pixel. 

I also needed to use overlap 60, because with the default overlap of 48, the axons had very sharp cuts : 
![Sharp cut](https://github.com/axondeepseg/axondeepseg/assets/119101905/606c7a1b-a8ef-4470-ad03-9c881804594f)

How can we use less patches ? With lower zoom factor ? 

I know my CPU is pretty weak but this is the best I've got, I just hope to speed things up so that using AxondeepSeg is faster or better than the current software.

**Reply by mathieuboudreau** [reply]
>the distance between 0 and 10 was 100µm

>Since: 0.1mm <=> 1040.0692pixels
>Then 1 pixel <=> 0.00009614mm

This works out for me so far.

>So when using AxonDeepSeg, I specify my pixel size as 0,96µm

This doesn't sound right to me.  0.00009614mm * (1 um / 1000 mm) = 0.09614 um. So this factor of 10 may be the reason that the software was resampling your image in such a way that increased the patch numbers. So first, I'd try processing it using this resolution. If there is still a very high number of patches (>30), then you changing the Zoom factor can further help in reducing it, which will speed up your processing.

As for the overlap edges, this was likely due to your mis-calculated resolution, as a high number of patches will only see a lot of axons "partially" (i.e. cutoff). Lowering the number of patch sizes will help with this, so you could try bringing it back down to 48. If the problem persists, we have a "no patch" option, but this should only be used on CPU if you had a small number of patches.

**Comment by hermancollin** [comment]
> > So when using AxonDeepSeg, I specify my pixel size as 0,96µm
> 
> This doesn't sound right to me. 0.00009614mm * (1 um / 1000 mm) = 0.09614 um. So this factor of 10 may be the reason that the software was resampling your image in such a way that increased the patch numbers.

To be fair if the resampling was off by a factor of 10 there is no way he would have obtained a good segmentation like the one he showed. In his original comment he specified:

```axondeepseg -t BF -i C:\Users\NFT10\Documents\NewMorpho\ID8.tif -s 0.096 --overlap 60 -r 1.5 3.0 -l 5```

We can confirm this in his screenshot. Maybe this was a typo. 

But as @mathieuboudreau, the easiest way to debug this problem would be to have access to an example image. @Debo12345678 you can share it with us at axondeepseg@googlegroups.com if you don't want to post it here.

**Reply by hermancollin** [reply]
Also, @Debo12345678, since you say your image is 3418x3404, I do have a recommendation. Usually, when I deal with large files, I crop a small part (let's say 500x500) to find the best options (e.g. running a zoom factor sweep like you did). This exploration process will be much easier for your computer on a smaller image, and when you find the best parameters, you can apply them on the large source image.

**Comment by Debo12345678** [comment, marked as answer]
Hello, I'm really sorry, I made a typing error in my message as suggested by @hermancollin
. My segmentations were indeed made with a pixel size of 0.096.
Thank you for your recommendations. 

With the default settings (overlap at 48 and no zoom) for my entire image, processing time is 1minute30. And 30 axial patches of shape [512, 512] were loaded. This is already a clear reduction in processing time compared to what I had previously. I had misunderstood how the overlap worked. 

I started by cropping a small part of my image (exactly 500x500pixel), and tested different zooms as you said, which took about 1m32 (4 axial patches are loaded for the 1.20 and 1.26 zoom and 9 axial patches are loaded for the 1.32 the 1.38 and 1.44 zooms). 

The 1.2 zoom seems to correspond to the best segmentation of the cut image, even if some sharp cuts remain. I run the segmentation of my whole image with a zoom of 1.2 and, out of curiosity, an overlap of 40, but it doesn't change anything: 42 axial patches are loaded and sharp cuts are still present: 
![screenshot_ID8_overlap40_zoom1 2](https://github.com/axondeepseg/axondeepseg/assets/119101905/54389ef5-a92e-4a17-8769-57103c4b00ff)


I run the full image with a zoom of 1.2 and an overlap of 30, and the sharp cuts decrease, but holes in some axons appear, so I mustn't go below 30. 
![screenshot_ID8_overlap30_zoom1 2](https://github.com/axondeepseg/axondeepseg/assets/119101905/f4912e21-2f63-4b1d-9654-6dfc988a6141)

My tests are a trial of AxonDeepSeg, my objective in the coming months will be to train a model myself (perhaps with a new computer). Thank you for your prompt replies

**Reply by hermancollin** [reply]
> The 1.2 zoom seems to correspond to the best segmentation of the cut image, even if some sharp cuts remain. I run the segmentation of my whole image with a zoom of 1.2 and, out of curiosity, an overlap of 40, but it doesn't change anything: 42 axial patches are loaded and sharp cuts are still present

> I run the full image with a zoom of 1.2 and an overlap of 30, and the sharp cuts decrease, but holes in some axons appear, so I mustn't go below 30

@Debo12345678 For this specific artifact, I would advise you to use the "no-patch" option as mentioned earlier. This kind of error only happens with patches.

> My tests are a trial of AxonDeepSeg, my objective in the coming months will be to train a model myself (perhaps with a new computer). Thank you for your prompt replies

Would you mind expanding on why you want to train your own model as opposed to using our pre-trained ones? I'm curious because the segmentations you showed already look pretty promising.
